v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/postgresql/PostgreSQLDialectAdapter.java

```diff
@@ -0,0 +1,495 @@
+          dataSource,
+          sourceSchemaReference,
+          tables,
+          e);
+      schemaDiscoveryErrors.inc();
+      throw new RetriableSchemaDiscoveryException(e);
+    } catch (SQLNonTransientConnectionException e) {
+      logger.error(
+          "Non Transient connection error while discovering table indexes for datasource={}, db={} tables={}",
+          dataSource,
+          sourceSchemaReference,
+          tables,
+          e);
+      schemaDiscoveryErrors.inc();
+      throw new SchemaDiscoveryException(e);
+    } catch (SQLException e) {
+      logger.error(
+          "Sql exception while discovering table schema for datasource={} db={} tables={}",
+          dataSource,
+          sourceSchemaReference,
+          tables,
+          e);
+      schemaDiscoveryErrors.inc();
+      throw new SchemaDiscoveryException(e);
+    } catch (SchemaDiscoveryException e) {
+      // Already logged.
+      schemaDiscoveryErrors.inc();
+      throw e;
+    }
+    ImmutableMap<String, ImmutableList<SourceColumnIndexInfo>> tableIndexes =
+        tableIndexesBuilder.build();
+    logger.info(
+        "Discovered Indexes for DataSource: {}, SourceSchemaReference: {}, Tables: {}.\nIndexes: {}",
+        dataSource,
+        sourceSchemaReference,
+        tables,
+        tableIndexes);
+    return tableIndexes;
+  }
+
+  /**
+   * Get query for the prepared statement to read columns within a range.
+   *
+   * @param tableName name of the table to read
+   * @param partitionColumns partition columns.
+   * @return Query Statement.
+   */
+  @Override
+  public String getReadQuery(String tableName, ImmutableList<String> partitionColumns) {
+    return addWhereClause("SELECT * FROM " + tableName, partitionColumns);
+  }
+
+  /**
+   * Get query for the prepared statement to count a given range.
+   *
+   * @param tableName name of the table to read.
+   * @param partitionColumns partition columns.
+   * @param timeoutMillis timeout of the count query in milliseconds. Set to 0 to disable timeout.
+   * @return Query Statement.
+   */
+  @Override
+  public String getCountQuery(
+      String tableName, ImmutableList<String> partitionColumns, long timeoutMillis) {
+    return addWhereClause(
+        String.format(
+            "SET statement_timeout = %s; SELECT COUNT(*) FROM %s", timeoutMillis, tableName),
+        partitionColumns);
+  }
+
+  /**
+   * Get query for the prepared statement to get min and max of a given column, optionally in the
+   * context of a parent range.
+   *
+   * @param tableName name of the table to read.
+   * @param partitionColumns if not-empty, partition columns. Set empty for first column of
+   *     partitioning.
+   */
+  @Override
+  public String getBoundaryQuery(
+      String tableName, ImmutableList<String> partitionColumns, String colName) {
+    return addWhereClause(
+        String.format("SELECT MIN(%s), MAX(%s) FROM %s", colName, colName, tableName),
```

Comment:
Would it be possible to UT the queries (with mocks if it's not possible to test with in-memory db)?

Reply:
Could you point me to where we do this with MySQL, so I can base my test on that?

Reply:
For specific boundary query, we test the query string
[here](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/sourcedb-to-spanner/src/test/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/mysql/MysqlDialectAdapterTest.java)

For larger parts like index discovery, we test via the mocks.

Reply:
Added the query string test for the missing boundary query.

IT test should cover the query execution, but let me know if you want specific UTs.